### PR TITLE
ARROW-11929: [C++][Dataset][Compute] Promote expression to the compute namespace

### PR DESCRIPTION
**Pull request by bkietz**
Moves Expression and its test and benchmark into the compute/exec/ directory. I haven't introduced an exec namespace.

### CODE REVIEW

**Review by lidavidm — APPROVED** [review]
LGTM. One nit about something that was recently introduced (I assume it just got lost in the rebasing).

I'll hold off on merging datasets stuff until this is through so we're not stuck in an endless rebase cycle.

**Comment by lidavidm** [comment]
Ah, looks like clang-format/cmake-format are unhappy, and the CMake magic that makes the new header gets installed is missing.

**Comment by nealrichardson** [comment]
This just moves the code? There's no change in how or where expressions are evaluated yet, correct? I.e. they're still only for datasets here, they're just in a different namespace?

**Comment by bkietz** [comment]
TODO: update bindings

**Review by lidavidm — COMMENTED** [review]
Sorry, I didn't notice this got rebased yesterday (feel free to ping me). This looks good, just needs to be linted/formatted, and it looks like `compute/expression_benchmark.cc` needs to be fixed. The AppVeyor failure is unrelated.

**Review by nealrichardson — APPROVED** [review]
R changes LGTM

**Review by lidavidm — APPROVED** [review]
LGTM, thanks!

File: cpp/examples/arrow/dataset_documentation_example.cc [hunk 1]
```diff
@@ -20,9 +20,9 @@
 
 #include <arrow/api.h>
 #include <arrow/compute/cast.h>
+#include <arrow/compute/exec/expression.h>
```
[the remark this answers is not in the document]
Comment (bkietz) [reply cpp/examples/arrow/dataset_documentation_example.cc:23]: https://issues.apache.org/jira/browse/ARROW-12605

**Comment by ianmcook** [comment]
@bkietz This caused a failure in the `test-r-minimal-build` nightly test:
https://dev.azure.com/ursacomputing/crossbow/_build/results?buildId=4423&view=logs&j=0da5d1d9-276d-5173-c4c4-9d4d4ed14fdb&t=d9b15392-e4ce-5e4c-0c8c-b69645229181&l=952
Do you know off the top of your head how to resolve this? Thanks

**Comment by lidavidm** [comment]
@ianmcook I think we just need to add Expression to compute/type_fwd.h. I'll do this now.

**Pull request closed without merging**